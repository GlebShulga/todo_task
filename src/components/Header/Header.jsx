import React, { useEffect, useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import { useSelector, useDispatch } from "react-redux";
import { useHistory, useLocation } from "react-router-dom";
import { matchPath } from "react-router";
import {
  setIsFilterStatusDone,
  updateChosenCategory,
} from "../../redux/actions/category";
import {
  setSearchCriteria,
  setEditingTaskMode,
} from "../../redux/actions/task";
import "./Header.scss";

const Header = () => {
  const dispatch = useDispatch();
  const history = useHistory();
  const { isFilterStatusDone } = useSelector((s) => s.category);
  const { pathname } = useLocation();

  const categoryParams = matchPath(pathname, { path: "/:category" });
  const choosenCategoryTitle = categoryParams?.params.category;

  const searchParams = matchPath(pathname, {
    path: "/:category/search/:subString",
  });

  useEffect(() => {
    const filterParams = matchPath(pathname, {
      path: "/categories/showDone=true",
    })?.isExact;
    if (filterParams) {
      dispatch(setIsFilterStatusDone(true));
    }
  }, []);

  const [search, setSearch] = React.useState(
    searchParams?.params.subString ?? ""
  );

  const onChangeSearch = (event) => {
    setSearch(event.target.value);
  };

  const onChangeFilterStatusDone = (e) => {
    if (e.target.checked) {
      dispatch(setIsFilterStatusDone(true));
      history.push("/categories/showDone=true");
    } else {
      dispatch(setIsFilterStatusDone(false));
      history.push("/categories/showDone=false");
    }
  };

  const onClickReturnToStartPage = () => {
    history.push(`/`);
    dispatch(updateChosenCategory({}));
    dispatch(setEditingTaskMode(false));
    dispatch(setIsFilterStatusDone(false));
  };

  const onClickSearch = () => {
    if (search.length > 0) {
      dispatch(setSearchCriteria(search));
      history.push(`/${choosenCategoryTitle}/search/${search}`);
    }
  };

  return (
    <div className="header" data-test="headerComponent">
      <button
        className="header-title"
        data-testid="HeaderTitle"
        onClick={onClickReturnToStartPage}
      >
        To-Do List
      </button>

      <button
        className="header-task-table_button"
        onClick={() => {
          history.push(`/tasktable`);
        }}
        data-testid="LinkToTaskTablePageButton"
      >
        Task table button
      </button>

      <div className="header-filter_form">
        <div className="status">
          <input
            type="checkbox"
            id="statusDone"
            name="statusDone"
            value={true}
            onChange={onChangeFilterStatusDone}
            checked={isFilterStatusDone}
            data-testid="StatusDoneCheckBox"
          />
          <label htmlFor="statusDone">Show done</label>
        </div>
        <div className="header-form">
          <input
            type="text"
            placeholder="Search"
            name="search"
            className="header-form_input"
            value={search}
            onChange={onChangeSearch}
            data-testid="HeaderFormInput"
          />
          <button
            className="header-form_button"
            onClick={onClickSearch}
            data-testid="SearchButton"
          >
            <FontAwesomeIcon icon={faTimes} />
          </button>
        </div>
      </div>
    </div>
  );
};

export default React.memo(Header);
