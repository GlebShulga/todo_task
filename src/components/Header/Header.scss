.header {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin: 0.5rem 2rem;
}

.header-title {
  font-size: calc(20px + 1vmin);
}

a {
  color: black;
  text-decoration: none;
}
a:hover {
 text-decoration: underline

}

.header-form {
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  margin-left: 1rem;
}

.header-filter_form {
  display: flex;
  align-items: center;
}

.header-form_input {
  padding: 0.5rem;
  font-size: 0.8rem;
  border: thin solid grey;
  float: left;
  width: 80%;
  background: #f1f1f1;
}

.header-form_button {
  float: left;
  width: 20%;
  padding: 0.5rem;
  background: #2196f3;
  color: white;
  font-size: 0.8rem;
  border: thin solid grey;
  border-left: none;
  cursor: pointer;
  &:hover {
    background: #0b7dda;
  }
  &::after {
    content: "";
    clear: both;
    display: table;
  }
}
