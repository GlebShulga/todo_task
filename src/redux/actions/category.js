/* eslint-disable no-loop-func */
import axios from "axios";
import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
} from "../types/category";

export function fetchCategoryList() {
  return (dispatch, getState) => {
    const store = getState();
    const url = store.router.location.pathname?.substring(1);
    const complexUrl = url?.match("[^/](.*?)(?=/)");
    const categoryUrl = complexUrl !== null ? complexUrl[0] : url;
    return axios
      .get("/api/v1/category")
      .then(({ data }) => {
        let current = data.find((cat) => cat.categoryTitle === categoryUrl);
        let updatedCategoryList = data;
        while (current?.parentCategoryId) {
          updatedCategoryList = updatedCategoryList.map((cat) =>
            cat.categoryId === current.parentCategoryId
              ? { ...cat, isExpanded: true }
              : cat
          );
          current = updatedCategoryList.find(
            (cat) => cat.categoryId === current.parentCategoryId
          );
        }

        const rootList = updatedCategoryList.filter(
          (el) => el.parentCategoryId === null
        );
        const titleList = data.map((category) => category.categoryTitle);
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
        dispatch({
          type: GET_CATEGORIES_LIST,
          categoryList: updatedCategoryList,
        });
        dispatch({ type: GET_CATEGORIES_TITLE_LIST, titleList });
      })
      .catch((err) => console.log(err));
  };
}

export function addCategory(categoryTitle, parentCategoryId, lvl) {
  return (dispatch) => {
    axios({
      method: "post",
      url: "/api/v1/category/:categoryTitle",
      data: {
        categoryTitle,
        parentCategoryId,
        lvl,
      },
    })
      .then(({ data: categoryList }) => {
        const rootList = categoryList?.filter(
          (el) => el.parentCategoryId === null
        );
        dispatch({ type: ADD_CATEGORY, categoryList });
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
      })
      .catch((err) => console.log(err));
  };
}

export function renameCategory(categoryId, categoryTitle) {
  return (dispatch) => {
    axios({
      method: "patch",
      url: "/api/v1/category",
      data: {
        categoryId,
        categoryTitle,
      },
    })
      .then(({ data: categoryList }) => {
        dispatch({ type: RENAME_CATEGORY, categoryList });
      })
      .catch((err) => console.log(err));
  };
}

export function delCategory(categoryId) {
  return (dispatch) => {
    axios({
      method: "delete",
      url: "/api/v1/category",
      data: {
        categoryId,
      },
    }).then(({ data: categoryList }) => {
      const rootList = categoryList.filter(
        (el) => el.parentCategoryId === null
      );
      dispatch({ type: DEL_CATEGORY, categoryList });
      dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
    });
  };
}

export function updateChosenCategory(chosenCategory) {
  return { type: UPDATE_CHOSEN_CATEGORY, chosenCategory };
}

export function setIsFilterStatusDone(data) {
  return { type: SET_IS_FILTER_STATUS_DONE, data };
}

export function setIsCreateTaskModalOpen(data) {
  return { type: SET_IS_CREATE_TASK_MODAL, data };
}
