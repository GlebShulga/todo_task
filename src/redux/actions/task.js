import axios from "axios";
import {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_SEARCH_CRITERIA,
} from "../types/task";

export function fetchTaskList() {
  return (dispatch, getState) => {
    axios("/api/v1/task")
      .then(({ data }) => {
        const store = getState();
        const url = store.router.location.pathname.substring(1);
        const taskUrl = url?.match("(?<=/)(.*?)(?=/edit)");
        dispatch({ type: GET_TASK_LIST, taskList: data });
        if (url.includes("edit")) {
          const chosenTask = data.find((task) => task.title === taskUrl[0]);
          dispatch({ type: UPDATE_CHOSEN_TASK, chosenTask });
        }
      })
      .catch((err) => console.log(err));
  };
}

export function addTask(title, categoryId) {
  return (dispatch) => {
    axios({
      method: "post",
      url: "/api/v1/task",
      data: {
        title,
        categoryId,
      },
    })
      .then(({ data: taskList }) => {
        dispatch({ type: ADD_TASK, taskList });
        dispatch({ type: GET_TASK_LIST, taskList });
      })
      .catch((err) => console.log(err));
  };
}

export function patchTask(taskId, status, title, description, categoryId) {
  return (dispatch) => {
    axios({
      method: "patch",
      url: "/api/v1/task",
      data: {
        taskId,
        status,
        title,
        description,
        categoryId,
      },
    })
      .then(({ data: taskList }) => {
        dispatch({ type: PATCH_TASK, taskList });
      })
      .catch((err) => console.log(err));
  };
}

export function updateChosenTask(chosenTask) {
  return { type: UPDATE_CHOSEN_TASK, chosenTask };
}

export function setEditingTaskMode(data) {
  return { type: SET_EDITING_TASK_MODE, data };
}

export function setSearchCriteria(searchCriteria) {
  return { type: SET_SEARCH_CRITERIA, searchCriteria };
}

export function setNewCategoryIdForTask(categoryId) {
  return { type: SET_NEW_CATEGORY_FOR_TASK, categoryId };
}
