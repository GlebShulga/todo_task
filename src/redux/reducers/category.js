import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
} from "../types/category";

const initialState = {
  categoryList: [],
  rootCategoryList: [],
  categoryTitleList: [],
  chosenCategory: {},
  isFilterStatusDone: false,
  isCreateTaskModalOpen: false,
  isOpenTaskTable: false,
};

// eslint-disable-next-line import/no-anonymous-default-export
export default (state = initialState, action) => {
  switch (action.type) {
    case GET_CATEGORIES_LIST:
    case ADD_CATEGORY:
    case DEL_CATEGORY:
    case RENAME_CATEGORY: {
      return { ...state, categoryList: action.categoryList };
    }
    case GET_ROOT_CATEGORY_LIST: {
      return { ...state, rootCategoryList: action.rootList };
    }
    case GET_CATEGORIES_TITLE_LIST: {
      return { ...state, categoryTitleList: action.titleList };
    }
    case UPDATE_CHOSEN_CATEGORY: {
      return { ...state, chosenCategory: action.chosenCategory };
    }
    case SET_IS_FILTER_STATUS_DONE: {
      return { ...state, isFilterStatusDone: action.data };
    }
    case SET_IS_CREATE_TASK_MODAL: {
      return { ...state, isCreateTaskModalOpen: action.data };
    }
    case SET_IS_OPEN_TASK_TABLE: {
      return { ...state, isOpenTaskTable: action.data };
    }
    default:
      return state;
  }
};


