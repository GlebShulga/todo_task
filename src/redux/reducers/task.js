import {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_SEARCH_CRITERIA,
} from "../types/task";

const initialState = {
  taskList: [],
  chosenTask: {},
  searchCriteria: "",
  isEditingTaskMode: false,
  newCategoryIdForTask: "",
};

// eslint-disable-next-line import/no-anonymous-default-export
export default (state = initialState, action) => {
  switch (action.type) {
    case GET_TASK_LIST:
    case ADD_TASK:
    case PATCH_TASK: {
      return { ...state, taskList: action.taskList };
    }
    case UPDATE_CHOSEN_TASK: {
      return { ...state, chosenTask: action.chosenTask };
    }
    case SET_EDITING_TASK_MODE: {
      return { ...state, isEditingTaskMode: action.data };
    }
    case SET_SEARCH_CRITERIA: {
      return { ...state, searchCriteria: action.searchCriteria };
    }
    case SET_NEW_CATEGORY_FOR_TASK: {
      return { ...state, newCategoryIdForTask: action.categoryId };
    }
    default:
      return state;
  }
};


