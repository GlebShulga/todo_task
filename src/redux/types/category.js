const ADD_CATEGORY = "ADD_CATEGORY";
const GET_CATEGORIES_LIST = "GET_CATEGORIES_LIST";
const GET_CATEGORIES_TITLE_LIST = "GET_CATEGORIES_TITLE_LIST";
const GET_ROOT_CATEGORY_LIST = "GET_ROOT_CATEGORY_LIST";
const RENAME_CATEGORY = "RENAME_CATEGORY";
const DEL_CATEGORY = "DEL_CATEGORY";
const UPDATE_CHOSEN_CATEGORY = "UPDATE_CHOSEN_CATEGORY";
const SET_IS_FILTER_STATUS_DONE = "SET_IS_FILTER_STATUS_DONE";
const SET_IS_CREATE_TASK_MODAL = "SET_IS_CREATE_TASK_MODAL";
const SET_IS_OPEN_TASK_TABLE = "SET_IS_OPEN_TASK_TABLE";

module.exports = {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
};
