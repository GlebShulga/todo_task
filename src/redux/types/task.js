const ADD_TASK = "ADD_TASK";
const GET_TASK_LIST = "GET_TASK_LIST";
const PATCH_TASK = "PATCH_TASK";
const UPDATE_CHOSEN_TASK = "UPDATE_CHOSEN_TASK";
const SET_EDITING_TASK_MODE = "SET_EDITING_TASK_MODE ";
const SET_NEW_CATEGORY_FOR_TASK = "SET_NEW_CATEGORY_FOR_TASK";
const SET_IS_SEARCH = "SET_IS_SEARCH";
const SET_SEARCH_CRITERIA = "SET_SEARCH_CRITERIA";

module.exports = {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_IS_SEARCH,
  SET_SEARCH_CRITERIA,
};
