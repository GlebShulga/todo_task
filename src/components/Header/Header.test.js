import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { shallow } from "enzyme";
import thunk from "redux-thunk";
import { mockData } from "../../../tests/mockStore";
import { SET_SEARCH_CRITERIA } from "../../redux/types/task";
import reactRouter from "react-router";
import Header from "./Header";
import { SET_IS_FILTER_STATUS_DONE } from "../../redux/types/category";

const middlewares = [thunk];
let mockStore;
const mockStoreConf = configureStore(middlewares);
const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe("<Header/>", () => {
  beforeEach(() => {
    mockStore = mockStoreConf(mockData);
  });

  it("should render Header Component", () => {
    const wrapper = shallow(
      <Provider store={mockStore}>
        <Header />
      </Provider>
    );
    wrapper.update();
  });

  it("Header snapshot", () => {
    const component = renderer.create(
      <Router>
        <Provider store={mockStore}>
          <Header />
        </Provider>
      </Router>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should push to TaskTable component", () => {
    const { getByTestId } = render(
      <Router>
        <Provider store={mockStore}>
          <Header />
        </Provider>
      </Router>
    );

    fireEvent.click(getByTestId("LinkToTaskTablePageButton"));
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
    expect(mockHistoryPush).toHaveBeenCalledWith(`/tasktable`);
  });

  it("should apply search criteria with search text", () => {
    const myInitialState = "task";

    React.useState = jest.fn().mockReturnValue([myInitialState, {}]);

    const { getByTestId } = render(
      <Router>
        <Provider store={mockStore}>
          <Header />
        </Provider>
      </Router>
    );

    fireEvent.click(getByTestId("SearchButton"));

    expect(mockStore.getActions()).toEqual([
      {
        type: SET_SEARCH_CRITERIA,
        searchCriteria: "task",
      },
    ]);
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
  });

  it("should apply search criteria with filter", () => {
    const searchCriteria = "task";

    React.useState = jest.fn().mockReturnValue([searchCriteria, () => {}]);
    jest
      .spyOn(reactRouter, "matchPath")
      .mockImplementation((pathname, pathObj) => ({
        params: {
          category: "category1",
        },
        isExact: true,
      }));
    const Component = (
      <Router>
        <Provider store={mockStore}>
          <Header />
        </Provider>
      </Router>
    );
    const { getByTestId } = render(Component);
    fireEvent.click(getByTestId("StatusDoneCheckBox"));
    fireEvent.change(getByTestId("HeaderFormInput"), {
      target: { value: searchCriteria },
    });
    fireEvent.click(getByTestId("SearchButton"));
    expect(mockHistoryPush).toHaveBeenCalledWith("/categories/showDone=true");
    expect(mockHistoryPush).toHaveBeenCalledWith("/category1/search/task");
    expect(mockStore.getActions()).toEqual([
      { type: SET_IS_FILTER_STATUS_DONE, data: true },
      {
        data: true,
        type: SET_IS_FILTER_STATUS_DONE,
      },
      {
        searchCriteria: searchCriteria,
        type: SET_SEARCH_CRITERIA,
      },
    ]);
    expect(mockHistoryPush).toHaveBeenCalledTimes(2);
  });
  it("should navigate to back previous page", () => {
    React.useState = jest.fn().mockReturnValue(["", () => {}]);
    const { getByTestId } = render(
      <Router>
        <Provider store={mockStore}>
          <Header />
        </Provider>
      </Router>
    );
    fireEvent.click(getByTestId("HeaderTitle"));
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
    expect(mockHistoryPush).toHaveBeenCalledWith(`/`);
  });
});
