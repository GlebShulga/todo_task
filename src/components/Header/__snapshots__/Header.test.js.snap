// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Header snapshot 1`] = `
<div
  className="header"
  data-test="headerComponent"
>
  <button
    className="header-title"
    data-test="title"
    onClick={[Function]}
  >
    To-Do List
  </button>
  <button
    className="header-task-table_button"
    onClick={[Function]}
  >
    Task table button
  </button>
  <div
    className="header-filter_form"
  >
    <div
      className="status"
    >
      <input
        checked={false}
        id="statusDone"
        name="statusDone"
        onChange={[Function]}
        type="checkbox"
        value={true}
      />
      <label
        htmlFor="statusDone"
      >
        Show done
      </label>
    </div>
    <div
      className="header-form"
    >
      <input
        className="header-form_input"
        name="search"
        onChange={[Function]}
        placeholder="Search"
        type="text"
        value=""
      />
      <button
        className="header-form_button"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
